File: solutions/haskell/104-accounts-merge.hs
{-
  Problem 104: Accounts Merge (LeetCode 721)
  Difficulty: Med
  Language: Haskell
-}
import qualified Data.Map as M
import Data.List (sort)

type UF = M.Map String String

find' :: UF -> String -> (UF, String)
find' uf x = case M.lookup x uf of
  Nothing -> (M.insert x x uf, x)
  Just p | p == x -> (uf, x)
         | otherwise -> let (uf', r) = find' uf p in (M.insert x r uf', r)

union' :: UF -> String -> String -> UF
union' uf a b = let (u1, ra) = find' uf a; (u2, rb) = find' u1 b
                in if ra == rb then u2 else M.insert ra rb u2

accountsMerge :: [[String]] -> [[String]]
accountsMerge accs =
  let (uf, owner) = foldl (\(u, o) (name:emails) ->
        foldl (\(u', o') e -> (union' u' (head emails) e, M.insert e name o'))
              (u, o) emails) (M.empty, M.empty) accs
      groups = foldl (\g e -> let (_, r) = find' uf e in
        M.insertWith (++) r [e] g) M.empty (M.keys owner)
  in map (\es -> (owner M.! head es) : sort es) (M.elems groups)

File: solutions/clojure/104-accounts-merge.clj
;; Problem 104: Accounts Merge (LeetCode 721)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn accounts-merge [accounts]
  (letfn [(find [uf x]
            (let [p (get uf x x)]
              (if (= p x) [uf x]
                (let [[uf' r] (find uf p)] [(assoc uf' x r) r]))))
          (union [uf a b]
            (let [[u1 ra] (find uf a) [u2 rb] (find u1 b)]
              (if (= ra rb) u2 (assoc u2 ra rb))))]
    (let [[uf owner] (reduce (fn [[u o] [name & emails]]
                               (reduce (fn [[u2 o2] e]
                                         [(union u2 (first emails) e) (assoc o2 e name)])
                                       [u o] emails))
                             [{} {}] accounts)
          groups (reduce (fn [g e]
                           (let [[_ r] (find uf e)]
                             (update g r (fnil conj []) e)))
                         {} (keys owner))]
      (map (fn [es] (cons (owner (first es)) (sort es))) (vals groups)))))
